eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -610,21 +619,85 @@ decodeCostModel lang = do
 getEvaluationContext :: CostModel -> PV1.EvaluationContext
 getEvaluationContext (CostModel _ _ ec) = ec
 
-newtype CostModels = CostModels {unCostModels :: Map Language CostModel}
+-- | See 'CostModels' for an explanation of how 'CostModelError' is used.
+--
+-- It would be preferable to store the 'PV1.CostModelApplyError' type itself,
+-- but there is currently no 'Eq' instance for it, which we require.
+data CostModelError = CostModelError String
+  deriving stock (Eq, Show, Ord, Generic)
+
+instance NoThunks CostModelError
+
+-- | For a known version of Plutus, attempting to construct a cost model with
+-- too few parameters (depending on the version) will result in an error.
+-- 'CostModelApplyError' exists to collect these errors in the 'CostModels' type.
+-- The 'CostModels' type itself needs to be flexible enough to accept any map
+-- of 'Word8' to '[Integer]', so that cost models can be placed in the protocol parameters
+-- ahead of changes to the Plutus evaluation context. In this way, serializing a cost model,
+-- updating software, and deserializing can result in errors going away.
+--
+-- Additionally, 'CostModels' needs to be able to store cost models for future version
+-- of Plutus, which we cannot yet even validate. These are stored in 'invalidCostModels'.
```

Comment:
Great and useful explanation! Would it be better positioned on top of `data CostModels` just below this line, rather than it is now on the `instance NFData` ? 
Not sure how the haddock looks if it's here

Reply:
:rofl: :facepalm: :rofl: 

good catch, thank you!